lxd/db/entity.go

```diff
@@ -276,60 +316,96 @@ func (c *Cluster) GetURIFromEntity(entityType int, entityID int) (string, error)
 			return "", fmt.Errorf("Failed to get operation: %w", err)
 		}
 
-		uri = fmt.Sprintf(cluster.EntityURIs[entityType], op.UUID)
-	case cluster.TypeStoragePool:
+		uri, err = entityType.URL("", "", op.UUID)
+		if err != nil {
+			return "", fmt.Errorf("Failed to get operation URL: %w", err)
+		}
+
+	case entity.TypeStoragePool:
 		_, pool, _, err := c.GetStoragePoolWithID(entityID)
 		if err != nil {
 			return "", fmt.Errorf("Failed to get storage pool: %w", err)
 		}
 
-		uri = fmt.Sprintf(cluster.EntityURIs[entityType], pool.Name)
-	case cluster.TypeStorageVolume:
-		var args StorageVolumeArgs
+		uri, err = entityType.URL("", "", pool.Name)
+		if err != nil {
+			return "", fmt.Errorf("Failed to get storage pool URL: %w", err)
+		}
 
+	case entity.TypeStorageVolume:
+		var args StorageVolumeArgs
+		var location string
 		err := c.Transaction(c.closingCtx, func(ctx context.Context, tx *ClusterTx) error {
 			args, err = tx.GetStoragePoolVolumeWithID(ctx, entityID)
 			if err != nil {
 				return err
 			}
 
+			if args.NodeID > 0 {
```

Comment:
Will this result in potentially a lot more queries?

Reply:
Ah I didn't realise that `tx.GetNodeWithID` would result in so many queries. We need the node name to use as the location of the storage volume, otherwise the volume may not be uniquely identified in a clustered set up.

Reply:
`tx.GetStoragePoolVolumeWithID` also performs a lot of unnecessary queries for what is required here (like getting the storage volume config).

Generally speaking this function is very inefficient.

Reply:
I'm primarily concerned with regressions firstly, I.e more queries than we do today. Could we load all cluster member names once at start of function call and then populate from a map in each iteration of the loop rather then potentially loading the same info repeatedly?

Reply:
Or is it only dealing with enriching a single record per call? 

Why wasn't  node name required before? Is this effectively a bug fix?

Reply:
Yes it's just one record per call. I guess it is a bug fix. I'll make it a separate commit at least. Then we can look at improving performance in https://github.com/canonical/lxd/issues/12868 perhaps?

Reply:
Ah ok not a bug fix per se. This method is only called in `UpsertWarning` to check if the entity ID is valid, e.g. it exists in the database. In that context it works as intended.

We would need this fix if we are to use it in any other context as the returned URL may refer to multiple volumes.

Reply:
I've added these changes as a separate commit.

Reply:
Ah ok thanks